explorer/interpreter/type_checker.cpp

```diff
@@ -2920,6 +2920,354 @@ auto TypeChecker::TypeCheckPattern(
+    assert(!row.empty());
+    std::vector<AbstractPattern> new_row;
+    switch (row[0].kind()) {
+      case AbstractPattern::Wildcard:
+        new_row.reserve(discriminator.size + row.size() - 1);
+        new_row.insert(new_row.end(), discriminator.size,
+                       AbstractPattern::MakeWildcard());
+        break;
+      case AbstractPattern::Compound: {
+        if (row[0].discriminator() != discriminator.discriminator) {
+          return std::nullopt;
+        }
+        assert(static_cast<int>(row[0].NumElements()) == discriminator.size);
+        new_row.reserve(discriminator.size + row.size() - 1);
+        row[0].AppendElementsTo(new_row);
+        break;
+      }
+      case AbstractPattern::Primitive:
+        // These cases should be rejected by the type checker.
+        llvm_unreachable("matched primitive against compound");
+    }
+    new_row.insert(new_row.end(), row.begin() + 1, row.end());
+    return std::move(new_row);
+  }
+
+  // Specialize the pattern matrix for the case that the first value matched
+  // uses `discriminator`, and its elements are matched.
+  auto Specialize(DiscriminatorInfo discriminator) const -> PatternMatrix {
+    PatternMatrix specialized;
+    for (auto& row : matrix_) {
+      // TODO: If we add support for "or" patterns, specialization might
+      // produce multiple rows here.
+      if (auto new_row = SpecializeRow(row, discriminator)) {
+        specialized.Add(std::move(new_row.value()));
+      }
+    }
+    return specialized;
+  }
+
+  // Specialize the pattern matrix for the case where the first value is known
+  // to be `value`, and is not matched.
+  auto Specialize(const Value& value) const -> PatternMatrix {
+    PatternMatrix specialized;
+    for (auto& row : matrix_) {
+      assert(!row.empty());
+      switch (row[0].kind()) {
+        case AbstractPattern::Wildcard:
+          break;
+        case AbstractPattern::Compound:
+          llvm_unreachable("matched compound against primitive");
+        case AbstractPattern::Primitive:
+          // TODO: Use an equality context here?
+          if (!ValueEqual(&row[0].value(), &value, std::nullopt)) {
+            continue;
+          }
+          break;
+      }
+      specialized.Add(std::vector<AbstractPattern>(row.begin() + 1, row.end()));
+    }
+    return specialized;
+  }
+
+  // Specialize the pattern matrix for the case where the first value uses a
+  // discriminator matching none of the non-wildcard patterns.
+  auto Default() const -> PatternMatrix {
+    PatternMatrix default_matrix;
+    for (auto& row : matrix_) {
+      assert(!row.empty());
+      switch (row[0].kind()) {
+        case AbstractPattern::Wildcard:
+          default_matrix.Add(
+              std::vector<AbstractPattern>(row.begin() + 1, row.end()));
+          break;
+        case AbstractPattern::Compound:
+        case AbstractPattern::Primitive:
+          break;
+      }
+    }
+    return default_matrix;
+  }
+
+ private:
```

Comment:
Repeating private again? Might be easier to read this with a forward declaration.

Reply:
Will fix repeated `private`.

I didn't follow what you would like to be forward-declared here, can you elaborate?

Reply:
I mean the higher level "separate class member declarations and definitions" so that it'd be easier to see the repeat `private` on a single screen. :)

Reply:
I'll do that when I move this to a separate file.

Reply:
👍

Reply:
Done.